db/snapshot-ctl.cc

```diff
@@ -104,10 +104,8 @@ future<> snapshot_ctl::do_take_column_family_snapshot(sstring ks_name, std::vect
     for (const auto& table_name : tables) {
         co_await _db.invoke_on_all([&] (replica::database &db) {
             auto& cf = db.find_column_family(ks_name, table_name);
-            // FIXME: need a better way.
-            // See https://github.com/scylladb/scylla/issues/10526
-            if (table_name.find(".") != sstring::npos) {
-                throw std::invalid_argument("Cannot take a snapshot of a secondary index by itself. Run snapshot on the table that owns the index.");
+            if (cf.schema()->is_view()) {
+                throw std::invalid_argument("Do not take a snapshot of a materialized view or a secondary index by itself. Run snapshot on the base table instead.");
```

Comment:
A question (as is often the case, the best way to answer such a question is with a test ;-)) - when you take a snapshot of a table with views, does it actually snapshot the views as well?

Reply:
No, it doesn't.  The idea, as far as I understand it, is that one can always re-generate MV/SI from the snapshot of the base table. Also, it is quite complicated to take a _coherent_ snapshot of both the base table and its view so one could restore both and hope they'll be in sync - hence it's better to regenerate MV/SI from the base table if needed rather than to go through the extra hoop of doing a distributed, coherent snapshot of both the base table and its view.